ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -843,261 +910,150 @@ mod tests {
-                vec_y_left.push(y_left);
-                vec_y_right.push(y_right);
-                vec_prss_left.push(prss_left);
-                vec_prss_right.push(prss_right);
-                vec_z_right.push(z_right);
-            }
+            let vec_x_left = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_x_right = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_y_left = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_y_right = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_prss_left = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_prss_right = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+
+            // compute z
+            let vec_z_left = vec_x_left.clone() & vec_y_left.clone()
+                ^ (vec_x_left.clone() & vec_y_right.clone())
+                ^ (vec_x_right.clone() & vec_y_left.clone())
+                ^ vec_prss_left.clone()
+                ^ vec_prss_right.clone();
+
+            // vector for unchecked elements
+            // i.e. these are used to fill the segments of the verifier and prover that are not part
+            // of this povers proof
+            let vec_z_right = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_x = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_y = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_prss = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
+            let vec_z = (0..1024).map(|_| rng.gen::<u8>()).collect::<BitVec<u8>>();
 
             // generate and push segments
             for i in 0..1024 / segment_size {
-                // conv to BitVec
-                let x_left =
-                    BitVec::<u8>::from_slice(&vec_x_left[i * segment_size..(i + 1) * segment_size]);
-                let x_right = BitVec::<u8>::from_slice(
-                    &vec_x_right[i * segment_size..(i + 1) * segment_size],
-                );
-                let y_left =
-                    BitVec::<u8>::from_slice(&vec_y_left[i * segment_size..(i + 1) * segment_size]);
-                let y_right = BitVec::<u8>::from_slice(
-                    &vec_y_right[i * segment_size..(i + 1) * segment_size],
-                );
-                let prss_left = BitVec::<u8>::from_slice(
-                    &vec_prss_left[i * segment_size..(i + 1) * segment_size],
-                );
-                let prss_right = BitVec::<u8>::from_slice(
-                    &vec_prss_right[i * segment_size..(i + 1) * segment_size],
+                // prover
+                // generate segments
+                let segment_prover = Segment::new(
+                    SegmentEntry::new(
+                        &vec_x_left[i * 8 * segment_size..(i + 1) * 8 * segment_size],
+                    ),
```

Comment:
OK, I think this is another example of the code being too complex with too many different classes. I do not think we need a `Segment` and `SegmentEntry` class. This seems like an area that should be simplified away with fewer types. This instantiation code just points to this very clearly in my mind.

Reply:
Segment and SegmentEntry are simple wrapper around BitSlices. Here we only need to explicitly use them since it is a test. Segment is convenient to use, otherwise I would need to list all 7 values (x_left ,...) as function inputs explicitly.